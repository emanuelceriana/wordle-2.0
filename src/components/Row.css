.row {
  display: flex;
  justify-content: center;
  flex-direction: row;
  align-items: center;
}

.row:focus {
  outline: none;
  box-shadow: none;
}
