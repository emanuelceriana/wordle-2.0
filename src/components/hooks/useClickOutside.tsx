import React, { useCallback, useEffect } from "react";

interface ClickOutsideProps {
  rowRefs: React.MutableRefObject<HTMLDivElement[]>;
  activeRowIdx: number;
}

const useClickOutside = ({ rowRefs, activeRowIdx }: ClickOutsideProps) => {
  const handleClickOutside = useCallback(
    (event: MouseEvent) => {
      const targetElement = event.target as Node;
      if (!rowRefs?.current[activeRowIdx]?.contains(targetElement)) {
        setTimeout(() => rowRefs?.current[activeRowIdx]?.focus(), 0);
      }
    },
    [rowRefs, activeRowIdx]
  );

  useEffect(() => {
    document.addEventListener("mousedown", handleClickOutside);
    return () => {
      document.removeEventListener("mousedown", handleClickOutside);
    };
  }, [handleClickOutside]);
};

export default useClickOutside;
