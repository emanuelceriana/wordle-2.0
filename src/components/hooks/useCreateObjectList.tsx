import { useEffect, useState } from "react";
import { v4 as uuidv4 } from "uuid";

interface ObjectList {
  id: string;
}

interface CreateObjectList {
  length: number;
}

export const useCreateObjectList = ({ length }: CreateObjectList) => {
  const [objectList, setObjectList] = useState<ObjectList[]>([]);

  useEffect(() => {
    const objectsArray: ObjectList[] = new Array(length)
      .fill(0)
      .map(() => ({ id: `grid-${uuidv4()}` }));
    setObjectList(objectsArray);
  }, [length]);

  return {
    objectList,
  };
};
