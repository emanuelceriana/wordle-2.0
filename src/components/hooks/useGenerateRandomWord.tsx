import { useCallback } from "react";
import wordsDS from "../../data/wordsDataSet.json";
import { QueryClient } from "react-query";
import { fetchRandomWord } from "../../api";

interface GetRandomWordProps {
  wordLength: number;
  queryClient: QueryClient;
}

export const useGenerateRandomWord = ({
  wordLength,
  queryClient,
}: GetRandomWordProps) => {
  const generate = useCallback(async () => {
    let possibleWords: string[] = [];

    for (const [key, words] of Object.entries(wordsDS)) {
      if (key === wordLength.toString()) {
        possibleWords = words;
      }
    }

    const randomIndex: number = Math.floor(
      Math.random() * possibleWords.length
    );

    try {
      const randomWord = await queryClient.fetchQuery("randomWord", () =>
        fetchRandomWord(wordLength)
      );
      if (randomWord && typeof randomWord === "string" && randomWord != "") {
        return randomWord;
      } else {
        throw new Error("Invalid Random Word");
      }
    } catch {
      return possibleWords[randomIndex];
    }
  }, [wordLength, queryClient]);

  return {
    generate,
  };
};
