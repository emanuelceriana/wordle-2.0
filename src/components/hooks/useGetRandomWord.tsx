import { useEffect, useState } from "react";
import wordsDS from "../../data/wordsDataSet.json";

interface GetRandomWordProps {
  wordLength: number;
}

export const useGetRandomWord = ({ wordLength }: GetRandomWordProps) => {
  const [randomWord, setRandomWord] = useState<string>();

  useEffect(() => {
    let possibleWords: string[] = [];

    for (const [key, words] of Object.entries(wordsDS)) {
      if (key === wordLength.toString()) {
        possibleWords = words;
      }
    }

    const randomIndex: number = Math.floor(
      Math.random() * possibleWords.length
    );
    setRandomWord(possibleWords[randomIndex]);
  }, [wordLength]);

  return {
    randomWord,
  };
};
