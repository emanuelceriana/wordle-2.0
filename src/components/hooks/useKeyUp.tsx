import {
  KeyboardEvent,
  KeyboardEventHandler,
  useCallback,
  useContext,
  useState,
} from "react";

import { AppContext, IAppContext } from "../../context/AppContext";
import { getAmountOfLetters, validateWord, ValidationArray } from "../../utils";

interface KeyUpProps {
  isRowActive: boolean;
  setFocusedInputIdx: (idx: number) => void;
  setActiveRowIdx: (idx: number) => void;
  activeRowIdx: number;
  inputValues: string[];
  setInputValues: (inputValues: string[]) => void;
  focusedInputIdx: number;
  wordLength: number;
  randomWord: string;
}

const hardModeMessageError = (
  <>
    <b>Hard Mode:</b> The word is not a valid word.
  </>
);

const useKeyUp = ({
  isRowActive,
  setFocusedInputIdx,
  setActiveRowIdx,
  activeRowIdx,
  inputValues,
  setInputValues,
  focusedInputIdx,
  wordLength,
  randomWord,
}: KeyUpProps) => {
  const [wordInputValidation, setWordInputValidation] = useState<
    ValidationArray[]
  >([]);
  const {
    playlist,
    userSettings,
    playSound,
    handleGlobalWordValidation,
    validateRealWord,
    setNotification,
  } = useContext<IAppContext>(AppContext);

  const validate = (randomWord: string, inputValues: string[]) => {
    handleGlobalWordValidation(validateWord(randomWord, inputValues));
    setWordInputValidation(validateWord(randomWord, inputValues));
  };

  const checkHardMode = useCallback(async () => {
    let next = true;
    if (userSettings.isHardModeActive) {
      const isValidRealWord = await validateRealWord(inputValues.join(""));
      if (!isValidRealWord) {
        setNotification(hardModeMessageError);
      }
      next = isValidRealWord;
    }
    return next;
  }, [inputValues, userSettings, validateRealWord, setNotification]);

  const handleKeyUp: KeyboardEventHandler<HTMLElement> = useCallback(
    async (e: KeyboardEvent) => {
      if (isRowActive) {
        const keyPressed = e.key.toLowerCase();
        if (
          keyPressed === "enter" &&
          getAmountOfLetters(inputValues) === wordLength
        ) {
          if (await checkHardMode()) {
            validate(randomWord, inputValues);
            setFocusedInputIdx(0);
            setActiveRowIdx(activeRowIdx + 1);
            playSound(playlist.Try);
          }
        } else if (keyPressed === "backspace") {
          if (
            inputValues[focusedInputIdx] &&
            inputValues[focusedInputIdx] !== ""
          ) {
            const updatedValues = [...inputValues];
            updatedValues[focusedInputIdx] = "";
            setInputValues(updatedValues);
          } else if (
            (!inputValues[focusedInputIdx] ||
              inputValues[focusedInputIdx] === "") &&
            focusedInputIdx > 0
          ) {
            setFocusedInputIdx(focusedInputIdx - 1);
          }
          playSound(playlist.Keyboard);
        } else if (/^[a-z]$/.test(keyPressed)) {
          const updatedValues = [...inputValues];
          updatedValues[focusedInputIdx] = keyPressed;
          setInputValues(updatedValues);
          playSound(playlist.Keyboard);
          if (focusedInputIdx < wordLength - 1) {
            setFocusedInputIdx(focusedInputIdx + 1);
          }
        }
      }
    },
    [
      isRowActive,
      inputValues,
      focusedInputIdx,
      userSettings,
      setActiveRowIdx,
      playSound,
      checkHardMode,
    ]
  );

  return {
    handleKeyUp,
    wordInputValidation,
  };
};

export default useKeyUp;
