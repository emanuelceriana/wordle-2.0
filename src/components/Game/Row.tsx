import { forwardRef, useContext, useMemo, useState, useEffect } from "react";
import { AppContext, IAppContext } from "../../context/AppContext";
import { LetterBlock } from "./LetterBlock";
import { useCreateObjectList } from "../hooks/useCreateObjectList";
import { useKeyUp } from "../hooks/useKeyUp";
import styles from "./Row.module.scss";
import { isCorrectWord } from "../../utils";

interface RowProps {
  idx: number;
  rowWord: string[];
  setRowsWord: (x: string[]) => void;
}

const defaultFocusedInputIdx = 0;

export const Row = forwardRef<HTMLDivElement, RowProps>(
  ({ idx, rowWord, setRowsWord }, rowRef) => {
    const {
      setActiveRowIdx,
      randomWord,
      wordLength,
      activeRowIdx,
      setIsWinner,
      setIsEndGame,
    } = useContext<IAppContext>(AppContext);

    const [focusedInputIdx, setFocusedInputIdx] = useState(
      defaultFocusedInputIdx
    );

    const isRowActive = useMemo(() => idx === activeRowIdx, [activeRowIdx]);

    const { objectList: rowInputs } = useCreateObjectList({
      length: wordLength,
    });

    const { handleKeyUp, wordInputValidation } = useKeyUp({
      isRowActive,
      setFocusedInputIdx,
      setActiveRowIdx,
      activeRowIdx,
      rowWord,
      setRowsWord,
      focusedInputIdx,
      wordLength,
      randomWord,
    });

    useEffect(() => {
      if (isCorrectWord(wordInputValidation)) {
        setIsWinner(true);
        setIsEndGame(true);
      }
    }, [wordInputValidation]);

    return (
      <div
        ref={rowRef}
        className={styles.row}
        tabIndex={-1}
        onKeyUp={handleKeyUp}
      >
        {wordLength &&
          rowInputs.map(({ id }, idx) => (
            <LetterBlock
              key={id}
              idx={idx}
              isFocused={focusedInputIdx === idx && isRowActive}
              value={rowWord?.[idx]}
              focusable={isRowActive}
              setFocusedInputIdx={setFocusedInputIdx}
              inputValidation={wordInputValidation[idx]}
            />
          ))}
      </div>
    );
  }
);
