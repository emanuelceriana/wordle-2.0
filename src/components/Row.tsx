import { forwardRef, useEffect, useMemo, useState } from "react";
import { LetterBlock } from "./LetterBlock";
import "./Row.css";
import { useCreateObjectList } from "./hooks/useCreateObjectList";
import useKeyDown from "./hooks/useKeyDown";
import { ValidationArray, isCorrectWord } from "./utils";

interface RowProps {
  idx: number;
  activeRowIdx: number;
  wordLength: number;
  randomWord: string;
  setActiveRowIdx: (idx: number) => void;
  setIsWinner: (value: boolean) => void;
  setIsEndGame: (value: boolean) => void;
  setGlobalWordArrayValidation: (value: ValidationArray[]) => void;
}

const defaultFocusedInputIdx = 0;

export interface RowRef {
  element: HTMLDivElement | null;
  handleKeyDown: (e: KeyboardEvent | React.KeyboardEvent<HTMLElement>) => void;
}

export const Row = forwardRef<HTMLDivElement, RowProps>(
  (
    {
      idx,
      activeRowIdx,
      wordLength,
      randomWord,
      setActiveRowIdx,
      setIsWinner,
      setIsEndGame,
      setGlobalWordArrayValidation,
    },
    rowRef
  ) => {
    const [focusedInputIdx, setFocusedInputIdx] = useState(
      defaultFocusedInputIdx
    );
    const [inputValues, setInputValues] = useState<string[]>([]);
    const isRowActive = useMemo(() => idx === activeRowIdx, [activeRowIdx]);

    const { objectList: rowInputs } = useCreateObjectList({
      length: wordLength,
    });

    const { handleKeyDown, wordArrayValidation } = useKeyDown({
      isRowActive,
      setFocusedInputIdx,
      setActiveRowIdx,
      activeRowIdx,
      inputValues,
      setInputValues,
      focusedInputIdx,
      wordLength,
      randomWord,
    });

    useEffect(() => {
      setGlobalWordArrayValidation(wordArrayValidation);
      if (isCorrectWord(wordArrayValidation)) {
        setIsWinner(true);
        setIsEndGame(true);
      }
    }, [wordArrayValidation]);

    return (
      <div ref={rowRef} className="row" tabIndex={-1} onKeyDown={handleKeyDown}>
        {wordLength &&
          rowInputs.map(({ id }, idx) => (
            <LetterBlock
              key={id}
              idx={idx}
              isFocused={focusedInputIdx === idx && isRowActive}
              value={inputValues[idx]}
              focusable={isRowActive}
              setFocusedInputIdx={setFocusedInputIdx}
              inputValidation={wordArrayValidation[idx]}
            />
          ))}
      </div>
    );
  }
);
