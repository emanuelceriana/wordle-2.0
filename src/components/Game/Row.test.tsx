import React from "react";
import { fireEvent, act } from "@testing-library/react";
import { Row } from "./Row";
import { renderWithContext, withContext } from "../../test-utils";
import * as useKeyUp from "../hooks/useKeyUp";

interface RowProps {
  idx: number;
  rowWord: string[];
  setRowsWord: () => void;
  ref: React.MutableRefObject<HTMLDivElement>;
}

interface MockValueContext {
  setActiveRowIdx: (x: number) => void;
  randomWord: string;
  wordLength: number;
  activeRowIdx: number;
  setIsWinner: (x: boolean) => void;
  setIsEndGame: (x: boolean) => void;
}

describe("<Row />", () => {
  const setRowsWord = jest.fn();
  const setActiveRowIdx = jest.fn();
  const setIsWinner = jest.fn();
  const setIsEndGame = jest.fn();

  let rowProps: RowProps = {
    idx: 0,
    rowWord: [],
    setRowsWord: setRowsWord,
    ref: { current: document.createElement("div") },
  };

  let mockValueContext: MockValueContext = {
    setActiveRowIdx,
    randomWord: "test",
    wordLength: 4,
    activeRowIdx: 0,
    setIsWinner,
    setIsEndGame,
  };

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it("should render without errors", () => {
    const { asFragment } = renderWithContext(
      <Row {...rowProps} />,
      mockValueContext
    );
    expect(asFragment()).toMatchSnapshot();
  });

  it("should render correctly the quantity of LetterBlocks depending on the wordLength value", () => {
    const wordLength = 8;
    mockValueContext.wordLength = wordLength;

    const { queryAllByTestId } = renderWithContext(
      <Row {...rowProps} />,
      mockValueContext
    );

    const letterBlocks = queryAllByTestId("letterBlock");

    expect(letterBlocks.length).toBe(wordLength);
  });

  it("should setIsWinner and setIsEndGame as true when the word is correct", () => {
    const ref = { current: document.createElement("div") };
    rowProps.ref = ref;

    const handleKeyUp = jest.fn();

    jest.spyOn(useKeyUp, "useKeyUp").mockImplementation(() => ({
      handleKeyUp,
      wordInputValidation: [],
    }));

    const { rerender } = renderWithContext(
      <Row {...rowProps} />,
      mockValueContext
    );

    act(() => {
      fireEvent.keyUp(ref.current, { key: "Enter", bubbles: true });
      expect(handleKeyUp).toHaveBeenCalledTimes(1);
      jest.spyOn(useKeyUp, "useKeyUp").mockImplementation(() => ({
        handleKeyUp,
        wordInputValidation: [
          { value: 1, ocurrences: 2, letter: "t" },
          { value: 1, ocurrences: 1, letter: "e" },
          { value: 1, ocurrences: 1, letter: "s" },
          { value: 1, ocurrences: 2, letter: "t" },
        ],
      }));
    });

    act(() => {
      rerender(withContext(<Row {...rowProps} />, mockValueContext));
    });

    expect(setIsWinner).toHaveBeenCalledTimes(1);
    expect(setIsEndGame).toHaveBeenCalledTimes(1);
  });
});
