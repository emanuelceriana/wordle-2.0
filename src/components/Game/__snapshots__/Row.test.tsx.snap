// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Row /> should render without errors 1`] = `
<DocumentFragment>
  <div
    class="row"
    tabindex="-1"
  >
    <div
      class="letterBlock focused"
      data-testid="letterBlock"
      tabindex="-1"
    />
    <div
      class="letterBlock"
      data-testid="letterBlock"
      tabindex="-1"
    />
    <div
      class="letterBlock"
      data-testid="letterBlock"
      tabindex="-1"
    />
    <div
      class="letterBlock"
      data-testid="letterBlock"
      tabindex="-1"
    />
  </div>
</DocumentFragment>
`;
